app/models/admin_dashboard_data.rb

```diff
@@ -3,17 +3,46 @@
 class AdminDashboardData
   include StatsCacheable
 
+  class Problem
+    VALID_PRIORITIES = ["low", "high"].freeze
+
+    attr_reader :message, :priority, :identifier
+
+    def initialize(message, priority: "low", identifier: nil)
+      @message = message
+      @priority = VALID_PRIORITIES.include?(priority) ? priority : "low"
+      @identifier = identifier
+    end
+
+    def to_s
+      @message
+    end
+
+    def to_h
+      { message: message, priority: priority, identifier: identifier }
+    end
+
+    def self.from_h(h)
+      h = h.with_indifferent_access
+      Problem.maybe_create(h[:message], priority: h[:priority], identifier: h[:identifier])
+    end
+
+    def self.maybe_create(message, priority: "low", identifier: nil)
+      return if message.blank?
```

Comment:
What do you think about having `initialize` check that `message` is not a blank string instead? I feel like that might be better because I don't think `Problem` works well with a blank string at all. For example, you wouldn't expect `Problem#to_s` to return `nil`.

Reply:
I guess the idea behind this is to allow for the checks and blocks to return nil, it's "syntactic sugar" more than anything. Is there a way to return nil from an `initialize` function if `message` is blank? The thing I tried didn't do that for me, maybe I just wasn't returning the right thing

Reply:
O I was going to suggest raising an error if `Problem#initialize` receives a blank string. Much easier to reason about the `Problem` class if `message` is never blank.

Reply:
The main reason I did it for things like this:

https://github.com/discourse/discourse/blob/54b27eb11cb3d29e8b53236a8b27699d2c26acce/app/models/admin_dashboard_data.rb#L56-L64

We are always compacting the problems received here so it doesn't matter if one returns `nil`, we don't really want to raise an error here. For example, say there are 3 `AdminDashboardData.problem_syms` that we are calling:

```ruby
# sym 1
public_send(sym)
=> "Some problem!"

# sym 2
public_send(sym)
=> nil # (no problem to report)

# sym 3
public_send(sym)
=> "Another problem!"
```

With this `maybe` method we end up with `[Problem("Some problem!"), nil, Problem("Another problem!")]` then we can just `compact` away the `nil`s rather than having to error, because we don't really _care_ that one of these returns nil, because that is valid.

Reply:
After looking at the the usage in the tests, I find `Problem.maybe_create` harder to reason about partly because `maybe_create` introduces a sense of uncertainty and I would want to check the return value from the method to make sure an instance of `Problem` is returned since there is chance the method might not return that per the method's name. 

```
 AdminDashboardData.problem_syms.each do |sym| 
    message = public_send(sym)
    problems << Problem.new(message) if message 
  end 
  
  AdminDashboardData.problem_blocks.each do |blk| 
    message = instance_exec(&blk)
    problems << Problem.new(message) if message 
  end 

  AdminDashboardData.problem_messages.each do |i18n_key|
    message = AdminDashboardData.problem_message_check(i18n_key)
    problems << Problem.new(message) if message 
  end
end
```

What do you think about something like the above?

Reply:
Yeah I can do that, better than raising an error at least :) I think it would be nice to also allow these to return actual `Problem` instances as well, which is what I am doing in the scheduled checks. Will make some changes :+1: